### PR TITLE
PR Authority lookup webcomponent

### LINKED ISSUE #57: Authority lookup webcomponent
## Goal
A webcomponent `geov-authority-lookup` which we can lookup URI entities from authority files (Wikidata and GND)

## Spec

Two webcomponents: `geov-authority-lookup` and `geov-authority-lookup-explorer`
This will use lobid (for gnd) and SPARQL (for Wikidata). See Confluence documentation.

### geov-authority-lookup-explorer

It's just a mini webcomponent which only a certain number of best occurences from a result of API can be display.

**Mockup**
Image
(The 'Open' button will be 'clear' and accompanied by an 'open' button)

**Features:**
With three buttons, we can copy, select or access the uri.

Select: The user can select an item. This emits the selected URI in an event on geov-authority-lookup.
Open: The user can open the website of the item in a new browser tab.
Copy: The user can copy the URI to the clipboard. (not available by default)

Inputs:

- api (e.g. 'wikidata' or 'gnd')
- keywords
- type (can be null)
- nbOccurencesMax (by default: 5) the number of occurences on display
- listButtons (by default ['select', 'open']) (other possible value in this list: 'copy')

### geov-authority-lookup

It's a webcomponent which the previous webcomponent can be used there. 
The webcomponent will mainly be completed with a search field, a type selection and a button.

**Mockup**
Image

On mobile
Image

Inputs
- list of type (**Person, Place, Group** to start)
- list of authority we want use (by default, 'gnd', 'wikidata')

## Examples of data with API

**Wikidata:**

https://query.wikidata.org/#SELECT%20%2a%20WHERE%20%7B%0A%20%20SERVICE%20wikibase%3Amwapi%20%7B%0A%20%20%20%20%20%20bd%3AserviceParam%20wikibase%3Aendpoint%20%22www.wikidata.org%22%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20wikibase%3Aapi%20%22EntitySearch%22%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20mwapi%3Asearch%20%22Kepler%22%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20mwapi%3Alanguage%20%22en%22.%0A%20%20%20%20%20%20%3Fitem%20wikibase%3AapiOutputItem%20mwapi%3Aitem.%0A%20%20%20%20%20%20%3Fnum%20wikibase%3AapiOrdinal%20true.%0A%20%20%7D%0A%20%20%3Fitem%20rdfs%3Alabel%20%3Fname%20filter%20%28lang%28%3Fname%29%20%3D%20%22en%22%29.%0A%20%20%3Fitem%20wdt%3AP31%20wd%3AQ5%0A%7D%20ORDER%20BY%20ASC%28%3Fnum%29%20LIMIT%2020

We can use item & name.

**GND (lobid)**
https://lobid.org/gnd/search?q=Johan,Kepler&filter=type:Person&format=json:preferredName

Returns (partial data):

```json

    "image": "https://commons.wikimedia.org/wiki/Special:FilePath/Portrait%20of%20Johannes%20Kepler.jpg?width=100",
    "label": "Kepler, Johannes",
    "id": "https://d-nb.info/gnd/118561448",
    "category": "Individualisierte Person"
```
We can use label & id.

**IdRef**
additional integration: idref
https://documentation.abes.fr/aideidrefdeveloppeur/index.html#UtiliserApiSolr

https://www.idref.fr/Sru/Solr?q=geogname_t:(Lyon)&start=0&rows=5&indent=on&sort=score%20desc&fl=id,ppn_z,affcourt_z&version=2.2&wt=json

We can use ppn_z (for concatenate URI, no alternative) & affcourt_z (label)

### PR DESCRIPTION
**Pull request by perrauda**
resolves #57

### CODE REVIEW

**Comment by perrauda** [comment]
@joschne can you review this ? thank you

**Comment by joschne** [comment]
Yes, I start to review now.

File: packages/design-system-web/src/components/geov-authority-lookup-explorer/lib/const.tsx [hunk 1]
```diff
@@ -39,3 +39,26 @@ export const TYPE_ALL_IDREF = 'all:';
...
+{
+  	(?item) text:query ('${keywords}') .
+  	?item rdfs:label ?name .
+  	?item a ?classUri .
+  	${type}
+}
+LIMIT ${pageSize}
+OFFSET 0`;
+
+export const TYPE_PERSON_GEOVISTORY = 'FILTER (?classUri = ontome:c21)';
```
Comment (joschne) [inline packages/design-system-web/src/components/geov-authority-lookup-explorer/lib/const.tsx:62]: why do you use a filter and not directly:

```
?item a ontome:c21
```

Sparql filter has to loop over rows while the other approach uses indexes and is much faster.

The overall solution could be:
```sparql
SELECT DISTINCT ?item ?name ?classUri
WHERE
{
  	(?item) text:query ('${keywords}') .
  	?item rdfs:label ?name .
  	?item a ${type} . 
}
LIMIT ${pageSize}
OFFSET 0`;

export const TYPE_PERSON_GEOVISTORY = 'ontome:c21';
...
```

File: packages/design-system-web/src/components/geov-authority-lookup-explorer/geov-authority-lookup-explorer.tsx [hunk 2]
```diff
@@ -169,6 +181,32 @@ export class GeovAuthorityLookupExplorer {
     }
   }
 
+  getGeovistoryData() {
+    this.title = TITLE_GEOVISTORY;
+    this.uriData = [];
+    const kw = this.keywords.trim().split(' ').map(word => `*${word}*`).join(' AND ');
```
Comment (joschne) [inline packages/design-system-web/src/components/geov-authority-lookup-explorer/geov-authority-lookup-explorer.tsx:187]: my prettier complains. Yours?

**Review by joschne — CHANGES_REQUESTED** [review]
I suggest some small changes. Else, good.

File: packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.tsx [hunk 3]
```diff
@@ -128,11 +146,10 @@ export class GeovAuthorityLookup {
           </ion-row>
         </ion-grid>
         <div class="containerResponsive">
-          <ion-grid class={this.classForceCol}>
-            <ion-row>
+          <div class={this.classForceCol}>
```
[the remark this answers is not in the document]
Comment (joschne) [reply packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.tsx:149]: Replace the line with this:

```tsx
 <div class="masonry-container" style={this.nbColMax ? { columnCount: this.nbColMax.toString() } : {}}>
```

File: packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.css [hunk 4]
```diff
@@ -6,26 +6,31 @@
   container-type: inline-size;
 }
 
-ion-grid {
-  --ion-grid-columns: 1;
+div.masonry-container{
+  column-count: 1;
+}
+
+div.masonry-item{
```
[the remark this answers is not in the document]
Comment (perrauda) [reply packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.css:13]: That's true. I've only just noticed it... I don't think I've seen it before... Thanks for that. Done

File: packages/design-system-web/src/components/geov-authority-lookup-explorer/geov-authority-lookup-explorer.tsx (hunk 2 above)
Comment (perrauda) [reply packages/design-system-web/src/components/geov-authority-lookup-explorer/geov-authority-lookup-explorer.tsx:187]: Done, with ESlint plugin (which not activated)

File: packages/design-system-web/src/components/geov-authority-lookup-explorer/lib/const.tsx (hunk 1 above)
Comment (perrauda) [reply packages/design-system-web/src/components/geov-authority-lookup-explorer/lib/const.tsx:62]: We need ?classUri and we need an usecase for no choice
But you're right, we can avoid FILTER(): ?item a ?classUri, ontome:c21 . 
Done

File: packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.css [hunk 5]
```diff
@@ -6,26 +6,31 @@
...
-  --ion-grid-columns: 2;
+div.forceCol2{
+  column-count: 2;
 }
 
-ion-grid.forceCol1 {
-  --ion-grid-columns: 1;
+div.forceCol1{
+  column-count: 1;
 }
```
[the remark this answers is not in the document]
Comment (perrauda) [reply packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.css:36]: Done

File: packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.tsx [hunk 6]
```diff
@@ -87,11 +105,11 @@ export class GeovAuthorityLookup {
...
     if (this.nbColMax == 2) {
-      this.classForceCol = 'forceCol2';
+      this.classForceCol = 'masonry-container forceCol2';
     } else if (this.nbColMax == 1) {
-      this.classForceCol = 'forceCol1';
+      this.classForceCol = 'masonry-container forceCol1';
     } else {
-      this.classForceCol = '';
+      this.classForceCol = 'masonry-container';
     }
```
[the remark this answers is not in the document]
Comment (perrauda) [reply packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.tsx:113]: Done

File: packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.tsx [hunk 7]
```diff
@@ -70,7 +88,7 @@ export class GeovAuthorityLookup {
 
   @State() type = '';
 
-  @State() classForceCol = '';
+  @State() classForceCol = 'masonry-container';
```
[the remark this answers is not in the document]
Comment (perrauda) [reply packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.tsx:91]: Done

File: packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (perrauda) [reply packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.tsx:149]: Done

File: packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.tsx [hunk 8]
```diff
@@ -128,11 +146,10 @@ export class GeovAuthorityLookup {
           </ion-row>
         </ion-grid>
         <div class="containerResponsive">
-          <ion-grid class={this.classForceCol}>
-            <ion-row>
+          <div class={this.classForceCol}>
               {this.keywords.length
```
[the remark this answers is not in the document]
Comment (perrauda) [reply packages/design-system-web/src/components/geov-authority-lookup/geov-authority-lookup.tsx:150]: Done with ESlint activation (but there still a problem with CRLF/LF but formatOnSave is ok)

**Review by joschne — APPROVED** [review]

**Pull request merged**